### PR TITLE
Implement simple array creation / serialization methods

### PR DESCRIPTION
**Pull request by SaharCarmel**

### CODE REVIEW

**Comment by kmaehashi** [comment]
Hi @SaharCarmel, ping me when ready to review 😃

**Comment by SaharCarmel** [comment]
I think that now I'm ready from you to review my changes.

File: cupy/_creation/from_data.py [hunk 1]
```diff
@@ -164,13 +174,67 @@ def fromfile(*args, **kwargs):
...
+    .. note::
+        Uses NumPy's ``fromstring`` and coerces the result to a CuPy array.
+
+    .. seealso:: :func:`numpy.fromstring`
+    """
+    return asarray(numpy.fromstring(*args, **kwargs))
 
 
-# TODO(okuta): Implement loadtxt
+def fromregex(*args, **kwargs):
```
Comment (kmaehashi) [inline cupy/_creation/from_data.py:210]: I'm sorry but `fromregex` cannot be implemented in CuPy as structured arrays are not supported yet in CuPy :bow:
Could you remove this one?

File: cupy/__init__.py [hunk 2]
```diff
@@ -240,6 +240,13 @@ def is_available():
 from cupy._creation.from_data import asarray  # NOQA
 from cupy._creation.from_data import ascontiguousarray  # NOQA
 from cupy._creation.from_data import fromfile  # NOQA
+from cupy._creation.from_data import fromfunction  # NOQA
+from cupy._creation.from_data import fromiter  # NOQA
+from cupy._creation.from_data import frombuffer  # NOQA
+from cupy._creation.from_data import fromstring  # NOQA
+from cupy._creation.from_data import loadtxt  # NOQA
+from cupy._creation.from_data import fromregex  # NOQA
```
Comment (kmaehashi) [inline cupy/__init__.py:248]:
```suggestion
```

File: docs/source/reference/creation.rst [hunk 3]
```diff
@@ -35,6 +35,13 @@ From existing data
    ascontiguousarray
    copy
    fromfile
+   fromregex
```
Comment (kmaehashi) [inline docs/source/reference/creation.rst:38]:
```suggestion
```

File: tests/cupy_tests/creation_tests/test_from_data.py [hunk 4]
```diff
@@ -497,6 +497,48 @@ def test_fromfile(self, xp):
...
+
+    @testing.numpy_cupy_array_equal()
+    def test_fromregex(self, xp):
+        with tempfile.TemporaryFile() as f:
+            f.write(b"\x00\x01\x02\x03\x04")
+            f.flush()
+            f.seek(0)
+            regexp = r"(\d+)\s+(...)"
+            return xp.fromregex(file=f, regexp=regexp,  dtype=numpy.int64)
+
```
Comment (kmaehashi) [inline tests/cupy_tests/creation_tests/test_from_data.py:525]:
```suggestion
```

File: tests/cupy_tests/io_tests/test_formatting.py [hunk 5]
```diff
@@ -18,3 +18,8 @@ def test_array_str(self):
         a = testing.shaped_arange((2, 3, 4), cupy)
         b = testing.shaped_arange((2, 3, 4), numpy)
         assert cupy.array_str(a) == numpy.array_str(b)
+
+    @testing.numpy_cupy_array_equal()
+    def test_array2string(self):
+        a = testing.shaped_arange((2, 3, 4), cupy)
+        return cupy.array2string(a)
```
Comment (kmaehashi) [inline tests/cupy_tests/io_tests/test_formatting.py:25]:
```suggestion
    def test_array2string(self):
        a = testing.shaped_arange((2, 3, 4), cupy)
        b = testing.shaped_arange((2, 3, 4), numpy)
        assert cupy.array2string(a) == numpy.array2string(b)
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: cupy/_io/text.py [hunk 6]
```diff
@@ -1,17 +1,59 @@
...
-
-# TODO(okuta): Implement loadtxt
-
-
-# TODO(okuta): Implement savetxt
+import numpy
+import cupy
+
+
+def savetxt(fname, X, fmt='%.18e', delimiter=' ', newline='\n', header='',
```
[the remark this answers is not in the document]
Comment (SaharCarmel) [reply cupy/_io/text.py:7]: What does ditton means?

File: cupy/_creation/from_data.py (hunk 1 above)
Comment (SaharCarmel) [reply cupy/_creation/from_data.py:210]: Sure, I think I have accidently implemented this one since there were a TODO there :sweat_smile:

**Review by SaharCarmel — COMMENTED** [review]
All done

File: cupy/_io/text.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kmaehashi) [reply cupy/_io/text.py]: Ah, just meant "the same comment as array2string" 😃  Confirmed it's fixed as expected!

File: cupy/_io/formatting.py [hunk 7]
```diff
@@ -42,3 +42,12 @@ def array_str(arr, max_line_width=None, precision=None, suppress_small=None):
     """
     return numpy.array_str(cupy.asnumpy(arr), max_line_width, precision,
                            suppress_small)
+
+
+def array2string(arr, *args, **kwargs):
```
Comment (kmaehashi) [inline cupy/_io/formatting.py:47]:
```suggestion
def array2string(a, *args, **kwargs):
```

File: cupy/_io/formatting.py [hunk 8]
```diff
@@ -42,3 +42,12 @@ def array_str(arr, max_line_width=None, precision=None, suppress_small=None):
...
+    """Return a string representation of an array.
+
+
+        .. seealso:: :func:`numpy.array2string`
```
Comment (kmaehashi) [inline cupy/_io/formatting.py:51]:
```suggestion
    .. seealso:: :func:`numpy.array2string`
```

File: cupy/_io/formatting.py [hunk 9]
```diff
@@ -42,3 +42,12 @@ def array_str(arr, max_line_width=None, precision=None, suppress_small=None):
...
+    """
+    return numpy.array2string(cupy.asnumpy(arr), *args, **kwargs)
```
Comment (kmaehashi) [inline cupy/_io/formatting.py:53]:
```suggestion
    return numpy.array2string(cupy.asnumpy(a), *args, **kwargs)
```

File: cupy/_creation/from_data.py [hunk 10]
```diff
@@ -135,7 +135,8 @@ def copy(a, order='K'):
     if fusion._is_fusing():
         if order != 'K':
             raise NotImplementedError(
-                'cupy.copy does not support `order` in fusion yet.')
+                'cupy.copy does not support `order` in fusion yet.'
+            )
```
Comment (kmaehashi) [inline cupy/_creation/from_data.py:139]:
```suggestion
                'cupy.copy does not support `order` in fusion yet.')
```

File: docs/source/reference/io.rst [hunk 11]
```diff
@@ -17,6 +17,16 @@ NumPy binary files (NPY, NPZ)
...
+
+.. autosummary::
+   :toctree: generated/
+
+   loadtxt
+   savetxt
+   genfromtxt
+   fromstring
 
 String formatting
```
Comment (kmaehashi) [inline docs/source/reference/io.rst:31]: `array2string` needs to be added to this section.
https://numpy.org/doc/stable/reference/routines.io.html

**Review by kmaehashi — COMMENTED** [review]
Thanks! Looks good to me except for nitpicks.

File: cupy/_creation/from_data.py (hunk 10 above)
Comment (SaharCarmel) [reply cupy/_creation/from_data.py:139]: Sure :laughing:

**Comment by SaharCarmel** [comment]
No nitpicks :package:

**Review by kmaehashi — APPROVED** [review]
LGTM!

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by SaharCarmel** [comment]
@kmaehashi how can I observe why the tests have failed?

File: tests/cupy_tests/creation_tests/test_from_data.py [hunk 12]
```diff
@@ -497,6 +497,39 @@ def test_fromfile(self, xp):
...
+    @testing.numpy_cupy_array_equal()
+    def test_genfromtxt(self, xp):
+        with tempfile.TemporaryFile() as fh:
+            fh.write(b"0 1\n2 3")
+            fh.flush()
+            fh.seek(0)
+            return xp.genfromtxt(fh, dtype="u1")
+
     def test_fromfile_big_endian(self, xp):
```
Comment (kmaehashi) [inline tests/cupy_tests/creation_tests/test_from_data.py:533]:
```suggestion
    @testing.numpy_cupy_array_equal()
    def test_fromfile_big_endian(self, xp):
```

**Comment by kmaehashi** [comment]
You can view the test log via `Details` link below.

**Comment by SaharCarmel** [comment]
It's kinda weird, I think that I have accidentally merged master into my fork and the little endian test crashes...

**Review by kmaehashi — APPROVED** [review]

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Review by kmaehashi — APPROVED** [review]

**Comment by kmaehashi** [comment]
LGTM! Thanks for the hard work @SaharCarmel!

**Pull request merged**